rclcpp/src/rclcpp/qos_event.cpp

```diff
@@ -69,17 +69,18 @@ QOSEventHandlerBase::is_ready(rcl_wait_set_t * wait_set)
 }
 
 void
-QOSEventHandlerBase::set_listener_callback(
-  rmw_listener_callback_t callback,
-  const void * user_data) const
+QOSEventHandlerBase::set_on_new_event_callback(
+  rcl_event_callback_t callback,
+  const void * user_data)
 {
-  rcl_ret_t ret = rcl_event_set_listener_callback(
+  rcl_ret_t ret = rcl_event_set_callback(
     &event_handle_,
     callback,
     user_data);
 
   if (RCL_RET_OK != ret) {
-    throw std::runtime_error("Couldn't set listener callback to QOSEventHandlerBase");
+    using rclcpp::exceptions::throw_from_rcl_error;
+    throw_from_rcl_error(ret, "failed to set the on new message callback for subscription");
```

Comment:
wrong error message?

Reply:
Our approach for un-setting the callbacks was passing `callback` and `user_data` as `nullptr`, where the RMW started counting the `unread_events` again, as when the callback is still not set like before adding a node to the executor. 
Now we get an error here,  where `rcl_event_set_callback` detects it is a `nullptr`, when the app is being finalized.
If passing `nullptr` is an error, we should have another way to "unset" the callbacks.

Reply:
Ok, I'll look at that.